File: app/src/main/java/space/pal/sig/view/apod/ApodViewModel.java
package space.pal.sig.view.apod;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import java.util.Date;
import java.util.List;

import space.pal.sig.model.Apod;
import space.pal.sig.repository.ApodRepository;

/**
 * SpaceNow
 * Created by Catalin on 7/24/2020
 **/
public class ApodViewModel extends AndroidViewModel {

    private final ApodRepository apodRepository;
    private final LiveData<Apod> todayApod;
    private final LiveData<List<Apod>> apodList;

    public ApodViewModel(@NonNull Application application, ApodRepository apodRepository) {
        super(application);
        this.apodRepository = apodRepository;
        todayApod = apodRepository.findByDate(new Date());
        apodList = apodRepository.findAllBeforeOfDateWithLimit(new Date(), 30);
    }

    public LiveData<Apod> getTodayApod() {
        return todayApod;
    }

    public LiveData<List<Apod>> getApodList() {
        return apodList;
    }
}

File: app/src/main/java/space/pal/sig/view/apod/ApodViewModel.kt
package space.pal.sig.view.apod

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.work.Data
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import androidx.work.WorkRequest
import space.pal.sig.model.dto.AstronomyPictureOfTheDayDto
import space.pal.sig.model.entity.AstronomyPictureOfTheDay
import space.pal.sig.repository.ApodRepository
import space.pal.sig.service.DownloadImageWorker
import space.pal.sig.util.DateExtensions.formatDate
import space.pal.sig.view.BaseViewModel
import java.util.*

/**
SpaceNow
Created by Catalin on 12/13/2020
 **/
class ApodViewModel(
        private val apodRepository: ApodRepository,
        application: Application
) : BaseViewModel(application) {

    companion object {
        const val IMAGE_URL = "image.url"
    }

    private val astronomyPictureOfTheDay = MediatorLiveData<AstronomyPictureOfTheDay>()
    private val workManager = WorkManager.getInstance(application)

    fun getAstronomyPictureOfTheDay(): LiveData<AstronomyPictureOfTheDay> {
        return astronomyPictureOfTheDay
    }

    fun submitDate(date: Date?) {
        date?.let {
            val liveApod = apodRepository.findByDate(date.formatDate())
            astronomyPictureOfTheDay.addSource(liveApod) { apod ->
                if (apod == null) {
                    apodRepository
                            .downloadByDate(date.formatDate())
                            .subscribe({
                                if (it.mediaType == AstronomyPictureOfTheDayDto.IMAGE) {
                                    apodRepository.save(it.toAstronomyPictureOfTheDay())
                                }
                            }, { it.printStackTrace() })
                } else {
                    astronomyPictureOfTheDay.value = apod
                }
            }
        }
    }

    fun downloadImage(url: String?) {
        val inputData = Data.Builder()
                .putString(IMAGE_URL, url)
                .build()
        val workRequest: WorkRequest = OneTimeWorkRequest
                .Builder(DownloadImageWorker::class.java)
                .setInputData(inputData)
                .build()
        workManager.enqueue(workRequest)
    }

}